### ISSUE TITLE
vw says "You are using label 3.40282e+38 not -1 or 1 as loss function expects!" on unlabeled examples

### ISSUE THREAD

**Issue by arielf** [JohnLangford/vowpal_wabbit#1111]
Scenario:
- Big data-set , about a billion points
- All unlabeled examples
- Loading a logistic regression model and trying to predict with it.

One in a while, very rare and out of the blue, in my case once every 350k examples, I get this message that doesn't make sense.  It seems like some overflow because the number refering to the label value in the message is MAXFLOAT to 6 significant digits.

But the message is wrong. The data set has no labels at all, let alone a MAXFLOAT label.

**Comment by arielf** [comment]
Narrowed this down to a very small example.
If you mix labeled and unlabeled examples and use `--bootstrap N` unlabeled examples seem to be initialized to max-float and error goes through the roof:

```
# -- mixed labeled and unlabeled data-set
$ cat 1.dat
 unlabeled-1| a b c
 unlabeled-2| d e
1 labeled-3| a b c
-1 labeled-4| d e
 unlabeled-5| d e

#
# -- Try to learn, unlabeled examples generate error as if label is initialized to maxfloat
#
$ vw -P 1 1.dat --loss_function logistic --bootstrap 2
Num weight bits = 18
learning rate = 0.5
initial_t = 0
power_t = 0.5
using no cache
Reading datafile = 1.dat
num sources = 1
average  since         example        example  current  current  current
loss     last          counter         weight    label  predict features
You are using label 3.40282e+38 not -1 or 1 as loss function expects!
0.693147 0.693147            1            1.0  unknown   0.0000        4
You are using label 3.40282e+38 not -1 or 1 as loss function expects!
0.693147 0.693147            2            2.0  unknown   0.0000        3
0.693147 0.693147            3            3.0   1.0000   0.0000        4
0.712379 0.770074            4            4.0  -1.0000   0.1484        3
You are using label 3.40282e+38 not -1 or 1 as loss function expects!
     inf      inf            5            5.0  unknown  -0.7724        3

finished run
number of examples per pass = 5
passes used = 1
weighted example sum = 5.000000
weighted label sum = 0.000000
average loss = inf
best constant = 0.000000
best constant's loss = 0.693147
total feature number = 17

#
# -- without loss_function logistic, -P/--progress stops working, and we see 'inf' loss values:
#
$ vw -P 1 1.dat  --bootstrap 2
Num weight bits = 18
learning rate = 0.5
initial_t = 0
power_t = 0.5
using no cache
Reading datafile = 1.dat
num sources = 1
average  since         example        example  current  current  current
loss     last          counter         weight    label  predict features
     inf      inf            1            1.0  unknown   0.0000        4
     inf      inf            2            2.0  unknown   0.0000        3
     inf 1.000000            3            3.0   1.0000   0.0000        4
     inf 1.216343            4            4.0  -1.0000   0.1029        3
     inf      inf            5            5.0  unknown  -0.6169        3

finished run
number of examples per pass = 5
passes used = 1
weighted example sum = 5.000000
weighted label sum = 0.000000
average loss = inf
best constant = 0.000000
best constant's loss = 1.000000
total feature number = 17
```

**Comment by JohnLangford** [comment]
Thanks, I believe this is fixed now.  In essence, we were not handling the error calculation properly in bootstrap when the example was unlabeled.

**Comment by arielf** [comment]
Confirming that this is fixed in latest. Thanks so much for the quick fix, John.
I plan to add a test case for it shortly.
